modules/currency-l0/src/main/scala/org/tessellation/currency/l0/snapshot/programs/Genesis.scala

```diff
@@ -58,30 +65,36 @@ object Genesis {
         .flatMap(OwnCollateralNotSatisfied.raiseError[F, Unit].unlessA)
 
       signedBinary <- stateChannelSnapshotService.createGenesisBinary(hashedGenesis.signed)
-      signedBinaryHash <- signedBinary.hashF
+      address = signedBinary.value.toAddress
+      _ <-
+        if (identifier =!= address) IdentifierNotEqualAddressFromGenesisDataError(identifier, address).raiseError[F, Unit]
+        else Applicative[F].unit
+      binaryHash <- signedBinary.toHashed.map(_.hash)
       _ <- stateChannelSnapshotClient.send(signedBinary)(globalL0Peer)
 
-      _ <- lastSignedBinaryHashStorage.set(signedBinaryHash)
+      _ <- lastBinaryHashStorage.set(binaryHash)
 
       signedIncrementalBinary <- stateChannelSnapshotService.createBinary(signedFirstIncrementalSnapshot)
-      signedIncrementalBinaryHash <- signedIncrementalBinary.hashF
+      incrementalBinaryHash <- signedIncrementalBinary.toHashed.map(_.hash)
       _ <- stateChannelSnapshotClient.send(signedIncrementalBinary)(globalL0Peer)
 
-      _ <- lastSignedBinaryHashStorage.set(signedIncrementalBinaryHash)
+      _ <- lastBinaryHashStorage.set(incrementalBinaryHash)
 
       _ <- consensusManager.startFacilitatingAfterRollback(
         signedFirstIncrementalSnapshot.ordinal,
         signedFirstIncrementalSnapshot,
         hashedGenesis.info
       )
-      _ <- logger.info(s"Genesis binary ${signedBinaryHash.show} and ${signedIncrementalBinaryHash.show} accepted and sent to Global L0")
+      _ <- logger.info(s"Genesis binary ${binaryHash.show} and ${incrementalBinaryHash.show} accepted and sent to Global L0")
     } yield ()
 
-    def accept(path: Path): F[Unit] =
+    def accept(path: Path, saltValue: Long): F[Unit] =
       genesisLoader
         .load(path)
         .map(_.map(a => (a.address, a.balance)).toMap)
-        .map(CurrencySnapshot.mkGenesis)
+        .map(CurrencySnapshot.mkGenesis(_, Salt(saltValue)))
         .flatMap(accept)
   }
+
+  case class IdentifierNotEqualAddressFromGenesisDataError(identifier: Address, genesisAddress: Address) extends NoStackTrace
```

Comment:
imo `IdentifierMismatch` or `IdentifierNotFromGenesisData` will be enough (you can always provide more descriptive error message if you want)

Reply:
Changed.